src/gui/notifications/dbusnotifier.cpp

```diff
@@ -0,0 +1,230 @@
+        return QUrl::fromLocalFile(h->rootPath());
+    }
+}
+
+void DBusNotifier::notificationClosed(uint id, uint /*reason*/)
+{
+    activeNotifications.remove(id);
+}
+
+void DBusNotifier::actionInvoked(uint id, const QString &action_key)
+{
+    if (!activeNotifications.contains(id)) {
+        return;
+    }
+    const Context& context = activeNotifications[id];
+    if (action_key == ACTION_NAME_DEFAULT) {
+        if (context.widget) {
+            context.widget->activateWindow();
+        }
+    }
+    else if (action_key == ACTION_NAME_OPEN_FINISHED_TORRENT) {
+        const BitTorrent::TorrentHandle* h = context.torrent;
+        // if there is only a single file in the torrent, we open it
+        if (h) {
+            QDesktopServices::openUrl(getUrlForTorrentOpen(h));
+        }
+    }
+}
+
+void DBusNotifier::showNotification(NotificationKind kind, const QString &title, const QString &message, const Context *context)
+{
+    QStringList actions;
+//    actions.append("default"); // This does not work with Plasma, it shows a button
+//    actions.append(ACTION_NAME_DEFAULT);
+    QVariantMap hints;
+    bool storeNotification = false;
+    int timeout = 0;
+    switch (kind)
+    {
+    case Notifier::TorrentFinished:
+        if (!capabilities.testFlag(BodyHyperlinks)) { // we need a button
+            actions.append(ACTION_NAME_OPEN_FINISHED_TORRENT);
+            actions.append(tr("Open", "Open donwloaded torrent"));
+            storeNotification = true;
+        }
+        hints[QLatin1String("category")] = QLatin1String("transfer.complete");
+        if (capabilities.testFlag(ActionIcons)) {
+            hints[QLatin1String("action-icons")] = true;
+        }
+        break;
+    case Notifier::TorrentIOError:
+        hints[QLatin1String("category")] = QLatin1String("transfer.error");
+        hints[QLatin1String("urgency")] = 2;
+        hints[QLatin1String("resident")] = true;
+        break;
+    case Notifier::UrlDownloadError:
+        hints[QLatin1String("category")] = QLatin1String("network.error");
+        hints[QLatin1String("urgency")] = 2;
+        hints[QLatin1String("resident")] = true;
+        break;
+    case Notifier::SearchFinished:
+         hints[QLatin1String("urgency")] = 0;
+         timeout = -1;
+         storeNotification = true;
+         break;
+    default:
+        break;
+    }
+    const QString APPLICATION_NAME = QLatin1String("qBittorrent");
+    hints[QLatin1String("desktop-entry")] = APPLICATION_NAME;
+
+    QDBusPendingReply<uint> reply = notificationsService->Notify(APPLICATION_NAME, 0, QLatin1String("qbittorrent"), // icon name
+                    title, message, actions, hints, timeout);
+    reply.waitForFinished();
+    if (storeNotification && reply.value() != 0) {
+        Context ctx = context ? *context : Context();
+        if (!ctx.widget) {
+            ctx.widget = mainWindow;
+        }
+        activeNotifications.insert(reply.value(), ctx);
```

Comment:
Don't store TorrentHandle pointer for future using. It's dangerous as TorrentHandle object can be deleted before you will use it and you will not be able to check it out. Either store QPointer<TorrentHandle> or store torrent hash and then call Session::findTorrent().